vscode/src/language-service/testExplorer.ts

```diff
@@ -0,0 +1,232 @@
+  const runHandler = (request: vscode.TestRunRequest) => {
+    if (!request.continuous) {
+      return startTestRun(request);
+    }
+  };
+
+  // runs an individual test run
+  // or test group (a test run where there are child tests)
+  const startTestRun = async (request: vscode.TestRunRequest) => {
+    // use the compiler worker to run the test in the interpreter
+
+    log.trace("Starting test run, request was", JSON.stringify(request));
+
+    const worker = getLocalCompilerWorker(context.extensionUri);
+
+    // request.include is an array of test cases to run, and it is only provided if a specific set of tests were selected.
+    if (request.include !== undefined) {
+      for (const testCase of request.include || []) {
+        await runTestCase(testController, testCase, request, worker);
+      }
+    } else {
+      // alternatively, if there is no include specified, we run all tests that are not in the exclude list
+      for (const [, testCase] of testController.items) {
+        if (request.exclude && request.exclude.includes(testCase)) {
+          continue;
+        }
+        await runTestCase(testController, testCase, request, worker);
+      }
+    }
+  };
+
+  /**
+   * Given a single test case, run it in the worker (which runs the interpreter) and report results back to the
+   * `TestController` as a side effect.
+   *
+   * This function manages its own event target for the results of the test run and uses the controller to render the output in the VS Code UI.
+   **/
+  async function runTestCase(
+    ctrl: vscode.TestController,
+    testCase: vscode.TestItem,
+    request: vscode.TestRunRequest,
+    worker: ICompilerWorker,
+  ): Promise<void> {
+    log.trace("Running Q# test: ", testCase.id);
+    if (testCase.children.size > 0) {
+      for (const childTestCase of testCase.children) {
+        await runTestCase(ctrl, childTestCase[1], request, worker);
+      }
+      return;
+    }
+    const run = ctrl.createTestRun(request);
+    const evtTarget = createDebugConsoleEventTarget((msg) => {
+      run.appendOutput(`${msg}\n`);
+    });
+    evtTarget.addEventListener("Result", (msg) => {
+      if (msg.detail.success) {
+        run.passed(testCase);
+      } else {
+        const failureLocation =
+          msg.detail?.value?.uri ||
+          (msg.detail?.value?.related &&
+            msg.detail.value.related[0].location?.source) ||
+          null;
+
+        const message: vscode.TestMessage = {
+          message: msg.detail.value.message,
+          location:
+            failureLocation === null
+              ? undefined
+              : {
+                  range: toVsCodeRange(msg.detail.value.range),
+                  uri: vscode.Uri.parse(failureLocation),
+                },
+        };
+        run.failed(testCase, message);
+      }
+      run.end();
+    });
+
+    const callableExpr = `${testCase.id}()`;
+    const uri = testMetadata.get(testCase)?.compilationUri;
```

Comment:
You don't really need to store `compilationUri` anywhere as you can use `testCase.uri` directly here, no?

Reply:
`testCase.uri` is the uri where the test case is located, but `compilationUri` is the uri of the compilation, i.e. manifest file...technically both would work but i thought using the compilation uri was cleaner

Reply:
I went ahead and switched to finding the compilation based on testCase.uri here